engine/pkg/election/storage.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"errors"
+	"time"
+)
+
+// ErrRecordConflict is returned when the record is conflicted.
+// It means that the record is not the latest in the storage.
+var ErrRecordConflict = errors.New("record conflict")
+
+// Storage describes the requirements of a storage that can be used for leader election.
+type Storage interface {
+	// Get gets the current record from the storage. It returns the current record stored
+	// in the storage and any error encountered. Note that if the record has not created yet,
+	// it should return an empty record and a nil error.
+	Get(ctx context.Context) (*Record, error)
+	// Update updates the record in the storage if the stored record version matches the given
+	// record version. It returns ErrRecordConflict if the stored record version does not match
+	// or any other error encountered.
+	Update(ctx context.Context, record *Record) error
+}
+
+// Record holds the information of a leader election.
+type Record struct {
+	// LeaderID is the ID of the leader. If it is empty, it means
+	// there is no leader and all members may try to become leader.
+	LeaderID string `json:"leader_id"`
+	// Members is the members that are eligible to become leader.
```

Comment:
`Members` also contains current leader if it exists, better to make the comment more clear.